keps/sig-api-machinery/20180415-crds-to-ga.md

```diff
@@ -143,29 +144,51 @@ The CRD `v1` API will be the same as the `v1beta1` but with all changes to the A
 
 ### Integration tests for GA
 
-TODO: complete this list
-
-* [Pruning for CustomResources KEP](https://github.com/kubernetes/enhancements/pull/709)
-  * TODO
-* [Vanilla OpenAPI Subset Design](https://docs.google.com/document/d/1pcGlbmw-2Y0JJs9hsYnSBXamgG9TfWtHY6eh80zSTd8)
-  * TODO
+* Ensure CRD themselves are v1<->v1beta1 round trippable
 * CRD versioning/conversion (https://github.com/kubernetes/kubernetes/issues/64136):
   * Ensure what is persisted in etcd matches the storage version
   * Set up a CRD, persist some data, changed the version, and ensure the previously persisted data is readable
   * Ensure discovery docs track a CRD through creation, version addition, version removal, and deletion
   * Ensure `spec.served` is respected
+* Tests for malformed inputs and errors (system should return sensible errors, not panic, and recover appropriately)
+  * TODO
 
 ### e2e tests for GA
 
-* Custom Resources should be readable and writable at all available versions (test for round trip-ability)
 * A Custom Resource watch should terminate if its CustomResourceDefinition is deleted or updated
 
 ### Scale Targets for GA
 
-* TODO quantify: Read/write latency of CRDs within X% of native Kubernetes types
-* TODO quantify: Latency degrades less than X% for up to 100k Custom Resources per CRD kind
-* TODO quantify: Webhook conversion QPS of a noop converter is within X% of QPS with no webhook
-* Coordinate with sig-scalability
+The scale target for GA of custom resources are defined by the same [API call latency
+SLIs/SLOs as the Kuberetes native types](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/api_call_latency.md#api-call-latency-slisslos-details):
+
+| SLI | SLO |
+| --- | --- |
+| Latency<sup>[1](#footnote1)</sup> of mutating<sup>[2](#footnote2)</sup> API calls for single objects for every (resource, verb) pair, measured as 99th percentile over last 5 minutes | In default Kubernetes installation, for every (resource, verb) pair, excluding virtual and aggregated resources and Custom Resource Definitions, 99th percentile per cluster-day <= 1s |
+| Latency<sup>[1](#footnote1)</sup> of non-streaming read-only<sup>[3](#footnote3)</sup> API calls for every (resource, scope<sup>[4](#footnote4)</sup>) pair, measured as 99th percentile over last 5 minutes | In default Kubernetes installation, for every (resource, scope) pair, excluding virtual and aggregated resources and Custom Resource Definitions, 99th percentile per cluster-day (a) <= 1s if `scope=resource` (b) <= 5s if `scope=namespace` (c) <= 30s if `scope=cluster` |
+
+And with these thresholds:
+
+| Quantity | Threshold scope=namespace | Threshold: scope=cluster |
+| --- | --- | --- |
+| Custom Resource objects per Definition | 1000 | 5000 |
+| Custom Resource Definitions | 100 | 500 |
```

Comment:
500 CRDs with validation schemas will lead to a enormous OpenAPI merge operation on each change of any of the CRD validation schemas. Do we take that into account?

Reply:
No, I need to do a scale run on this still. I'll do that this week and report back.

Reply:
ref https://github.com/kubernetes/kubernetes/pull/67205#discussion_r211183911. Implementation details changed from that time, but we have some insights. Before actual re-measuring, I imagine: 

- kube-aggregator's performance doesn't get affected by #CRDs. It pulls latest spec from apiextensions-apiserver every second
- each change of any of the CRD validation schemas triggers crd-openapi-controller to add all CRD specs up. The time grows linearly against the # of Paths and Definitions (O(#CRD)) (~0.4ms for in-tree specs (~450 paths))

Reply:
> "SLOs being satisfied when all dimensions are being pushed at the same time"

note that mutating CRD can lead to performance impact on its CR, every CRD updates will cascade create/destroy the CR storage, even tho it's working in a graceful and asynchronous way now https://github.com/kubernetes/kubernetes/pull/77816, but we should probably SLI/SLO for CR is only guaranteed when the CRD stays unmodified.

Reply:
@roycaihw have you verified that we are actually linear everywhere in the merging? We traverse the spec for each addition. The spec grows linearly => `O(n*O(n))=O(n*n)` complexity.

Reply:
I think for each schema update, in crd openapi controller, [the merging](https://github.com/kubernetes/kubernetes/blob/13af8f286de90cb3d287aab87631d548022c61ed/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/aggregator.go#L23-L70) traverse all the CRD specs linearly; in [kube-aggregator](https://github.com/kubernetes/kubernetes/blob/fc5ef37db177158b7f0afce5bb59908803193bf7/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L179) it's [harder to reason about](https://github.com/kubernetes/kubernetes/blob/fc5ef37db177158b7f0afce5bb59908803193bf7/vendor/k8s.io/kube-openapi/pkg/aggregator/aggregator.go#L157-L259) -- I don't think there is any conflict in CRDs, so approximately it should also be linear

I did some [experiment](https://gist.github.com/roycaihw/21150606b75fd43e207467cbaa81ac2a). The latency seems to be linearly increasing, but the numbers are much worse than I expected -- reaching ~30s when there are ~400 CRDs. Running the experiment in presubmit contains more noise, and also doesn't good https://github.com/kubernetes/kubernetes/pull/80192. I will try to get some breakdown numbers tomorrow

Reply:
I've penciled this in at 500 with TODO to verify it. Let me know once we have a firm number and I'll put it in.

Reply:
@roycaihw  I've added some comments to the doc about how this 500 limit is due to OpenAPI publishing latency and not the scalability SLIs/SLOs

Reply:
@jpbetz Agree we should document this. OpenAPI publishing is asynchronous so it doesn't fall under the read-latency SLIs/SLOs, but people should be aware of the numbers here.

I finally got the time to dig into what happened here-- within apiextensions-apiserver:
- building and merging specs for all CRDs are trivial as expected. [Merging a single spec](https://github.com/kubernetes/kubernetes/blob/ab3bf7237dde7cab13af1a8682bf8229345d8d49/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/aggregator.go#L51) took 5-20µs; merging 400 specs for CRD with non-trivial [schema](https://github.com/kubernetes/kubernetes/blob/ab3bf7237dde7cab13af1a8682bf8229345d8d49/test/e2e/apimachinery/crd_publish_openapi.go#L615-L660) took ~5ms
- what's expensive is [updating the spec in our OpenAPIService](https://github.com/kubernetes/kubernetes/blob/ab3bf7237dde7cab13af1a8682bf8229345d8d49/vendor/k8s.io/kube-openapi/pkg/handler/handler.go#L106-L139), which involves a few JSON (de)serialization. The time spent [processing one CRD in crd_openapi_controller](https://github.com/kubernetes/kubernetes/blob/ab3bf7237dde7cab13af1a8682bf8229345d8d49/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go#L143-L147) when there were 400 CRDs was 1.408189538s, while 1.401639874s of which was spent in [updating the spec in our OpenAPIService](https://github.com/kubernetes/kubernetes/blob/ab3bf7237dde7cab13af1a8682bf8229345d8d49/vendor/k8s.io/kube-openapi/pkg/handler/handler.go#L106-L139)

Summary: 
- within apiextensions-apiserver, the slow part is to serialize a large in-memory go-openapi spec to JSON and protobuf, which we serve from /openapi/v2
- **TODO**: I'm going to instrument the time elapsed in kube-aggregator

Reply:
same condition as above, when there are ~400 CRDs with non-trivial [schema](https://github.com/kubernetes/kubernetes/blob/ab3bf7237dde7cab13af1a8682bf8229345d8d49/test/e2e/apimachinery/crd_publish_openapi.go#L615-L660) in the cluster, changing one CRD schema will cause the following:
- kube-aggregator downloads latest (note: we do have etag support) openapi spec from apiextensions-apiserver [**in JSON**](https://github.com/kubernetes/kubernetes/blob/3c332a9487552ef8a778522b2b3a9ac9e3564287/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go#L62), which takes 2.299707048s
- trivial: kube-aggregator [merges CRD specs with built-in specs](https://github.com/kubernetes/kubernetes/blob/3c332a9487552ef8a778522b2b3a9ac9e3564287/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L171-L182)-- 32.939µs, 9.823802ms, 71.176473ms for each of the three delegation targets
- kube-aggregator [updates the spec in OpenAPIService](https://github.com/kubernetes/kubernetes/blob/ab3bf7237dde7cab13af1a8682bf8229345d8d49/vendor/k8s.io/kube-openapi/pkg/handler/handler.go#L106-L139), which takes 1.739409296s
- in total kube-aggregator spends 4.120326312s in [processing apiextensions-apiserver](https://github.com/kubernetes/kubernetes/blob/3c332a9487552ef8a778522b2b3a9ac9e3564287/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L110-L112)

Summary: 
- in kube-aggregator, most of the time was spent downloading the spec from apiextensions-apiserver in JSON, and also serializing a large in-memory go-openapi spec to JSON and protobuf
- latency for a spec change in apiextensions-apiserver to get reflected in kube-apiserver: 5~10s = 4s (stale update) + 1s (successfulUpdateDelayLocal) + 4s (real update)
- one improvement we can do: download openapi spec in protobuf from apiextensions-apiserver; it might not hold for custom aggregated apiserver

Reply:
Thanks for gathering all the data @roycaihw. I'll close this thread now. If you need to save any data from it for posterity, feel free to copy it into the scale numbers doc.